Improve -Ztime-passes

nnethercote opened the pull request:
Two improvements that make `-Ztime-passes` more useful.

r? @Zoxc

Zoxc commented on src/librustc_driver/lib.rs:
```diff
@@ -1179,8 +1184,13 @@ pub fn main() {
+    // `print_time_passes_entry` increments the depth to 0 before printing.
+    // The extra `\t` is necessary to align this label with the others.
+    set_time_depth(-1);
+    print_time_passes_entry(show_time, "\ttotal", start.elapsed());
```
You should change `print_time_passes_entry` to not modify the depth instead.

Zoxc replied:
You could also print the crate name with the total, that's useful if you're compiling multiple crates.

nnethercote replied:
I have changed `print_time_passes_entry` as you suggest.

I can't work out how to print the crate name. I see `Config::crate_name`, but it gets set to `None` initially and I can't see if/when it is updated.

Zoxc commented on src/librustc_driver/lib.rs:
```diff
@@ -126,12 +128,12 @@ pub fn run_compiler(
     file_loader: Option<Box<dyn FileLoader + Send + Sync>>,
     emitter: Option<Box<dyn Write + Send>>
-) -> interface::Result<()> {
+) -> interface::Result<bool> {
```
Instead of changing the return type here, use `Callbacks` to find out if timings are enabled.

nnethercote replied:
I don't understand this suggestion. We pass in a `DefaultCallbacks` instance, and that type implements the `Callback` trait via the default ops, which are no-ops. What would the code look like?

Zoxc replied:
Just create a new struct (with a bool field) and implement the `Callbacks` trait for it. Override the `config` method to read out if timings are enabled.

nnethercote replied:
Returning the bool is a lot simpler. No traits involved!

Zoxc replied:
It also doesn't make sense for `run_compiler` to return just a random `bool` though, and this also breaks RLS and Miri for  no reason.

Zoxc commented on src/librustc_mir/borrow_check/nll/region_infer/mod.rs:
```diff
@@ -404,22 +404,6 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         mir_def_id: DefId,
         errors_buffer: &mut Vec<Diagnostic>,
-    ) -> Option<ClosureRegionRequirements<'tcx>> {
-        common::time_ext(
```
Probably should just keep this and use `-Z time` instead? I don't think anyone will miss it though.

retep998 replied:
On the contrary, why should it even be there?

nnethercote replied:
Giving it some more thought, I'm going to go a step further and remove -Ztime. (Or perhaps leave it as a synonym for -Ztime-passes.) It's silly to have two timing flags, where one is useless (the current situation). It's also silly to have two timing flags that give almost-identical output (the situation with this patch).

Zoxc replied:
It makes sense to have a high level and a more detailed timing pass to me, but maybe the detailed stuff should just be left to `-Z self-profile`.

nnethercote replied:
Eh, I'll leave the `time-passes`/`time` split for now even though I don't like it. But I strongly believe that removing the `solve_nll_region_constraints` entries is the right decision; they are zero-value spam.

nnethercote commented:
@Zoxc: new code is up. It addresses some of your comments, but not all, as per the discussions above. I think it's a clear improvement over the status quo, see what you think.

nnethercote commented:
@Zoxc: what do you think about the latest code?

nnethercote commented:
@Zoxc: I updated the code to use the `Callbacks` trait instead of the `run_compiler` return value to propagate the `bool`.

Zoxc commented on src/librustc_driver/lib.rs:
```diff
@@ -114,9 +116,17 @@ pub trait Callbacks {
 }
 
-pub struct DefaultCallbacks;
+#[derive(Default)]
```
Create a private copy of this struct instead of changing the public one.

nnethercote replied:
Good suggestion. I have uploaded new code that does this.

Zoxc commented on src/librustc_driver/lib.rs:
```diff
@@ -1169,18 +1183,24 @@ pub fn init_rustc_env_logger() {
-    });
+    };
+    // The extra `\t` is necessary to align this label with the others.
+    set_time_depth(0);
```
Why is this `\t` necessary?

nnethercote replied:
The comment immediately above explains why. If the comment is not clear, please suggest an alternative wording.

Zoxc replied:
I don't understand the reason. The other calls have no extra `\t`, so why is it needed here?

nnethercote replied:
Sample output with the extra `\t` (with 8-space tabs):
```
  time: 0.000; rss: 101MB       codegen
  time: 0.000; rss: 101MB       serialize work products
  time: 0.000; rss: 101MB       linking
time: 0.049; rss: 77MB          total
```
Without the extra `\t`:
```
  time: 0.000; rss: 101MB       codegen
  time: 0.000; rss: 101MB       serialize work products
  time: 0.000; rss: 101MB       linking
time: 0.049; rss: 77MB  total
```
(The tabs were converted to spaces when I pasted them into this comment, but the layout is indicative.)

Zoxc replied:
That is kind of supposed to happen? =P

Zoxc commented on src/librustc_driver/lib.rs:
```diff
@@ -1169,18 +1183,24 @@ pub fn init_rustc_env_logger() {
+    };
+    // The extra `\t` is necessary to align this label with the others.
+    set_time_depth(0);
+    print_time_passes_entry(callbacks.time_passes, &format!("\ttotal"), start.elapsed());
```
No need for `format!` here.

nnethercote replied:
Yep, fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #61775) made this pull request unmergeable. Please resolve the merge conflicts.

nnethercote commented:
@Zoxc: I have rebased. As far as I can tell, I have addressed all your concerns and answered all your questions. I think it's ready to land.

Zoxc commented:
@bors r+

I think we could probably just remove the indentation support. It's broken for LLVM and for the parallel compiler anyway.

bors commented:
:pushpin: Commit 87b103d4a99193bf57a17866fb99596df7d31d8b has been approved by `Zoxc`

<!-- @bors r=Zoxc 87b103d4a99193bf57a17866fb99596df7d31d8b -->
<!-- homu: {"type":"Approved","sha":"87b103d4a99193bf57a17866fb99596df7d31d8b","approver":"Zoxc"} -->

nnethercote commented:
The indentation is crucial to understand which passes are sub-passes. I understand it's broken for the back-end, but even then it's still highly useful for the other passes.
